### PR TITLE
Add calculation types and more grouped query implementation

### PR DESCRIPTION
**Pull request by xichen2020**
cc @notbdu @black-adder 

This PR adds relevant types for calculation and adds more logic for grouped query implementation.

### CODE REVIEW

File: query/query.go [hunk 1]
```diff
@@ -301,11 +302,14 @@ func (q *RawQuery) validateCalculations() error {
...
+	// If the calculation operation is count, the field should be omitted,
+	// in which case the number of events containing the corresponding group key
+	// is returned. All other operations require the field be specified.
 	for _, calc := range q.Calculations {
-		// It's okay to omit the column if the calculation operation is count,
-		// in which case the number of events containing the corresponding group key
-		// is returned. All other operations require the field be specified.
-		if calc.Field == nil && calc.Op != Count {
+		if calc.Field != nil && !calc.Op.RequiresField() {
+			return fmt.Errorf("field specified %v for calculation op %v", *calc.Field, calc.Op)
```
Comment (black-adder) [inline query/query.go]: nit: maybe just me but an error that reads: `field %v CANNOT be specified for calculation op` makes it easier to understand why an error was thrown

File: calculation/calculation_op.go [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+		Max: field.ValueTypeSet{
+			field.IntType:    struct{}{},
+			field.DoubleType: struct{}{},
+			field.StringType: struct{}{},
+			field.TimeType:   struct{}{},
+		},
+	}
+)
+
+func init() {
```
[the remark this answers is not in the document]
Comment (xichen2020) [reply calculation/calculation_op.go]: Right yeah exactly.

File: query/query.go (hunk 1 above)
Comment (xichen2020) [reply query/query.go]: Fair enough, will update.

File: query/grouped_result.go [hunk 3]
```diff
@@ -1,49 +1,36 @@
...
 type GroupedResults struct {
-	OrderBy            []OrderBy
-	Limit              int
-	ValuesLessThanFn   field.ValuesLessThanFn
-	RequiredFieldPaths [][]string
-
-	// If `OrderBy` is not empty, the groups are sorted in the order dictated by the `OrderBy`
-	// clause in the query.
-	Groups []ResultGroup
+	GroupBy                     [][]string
```
[the remark this answers is not in the document]
Comment (xichen2020) [reply query/grouped_result.go]: Correct, will leave a comment. Not a bad idea on the type alias but I'm also a bit worried about the cognitive overhead, so will punt on the type alias a bit and can add if this becomes an issue.

**Pull request merged**